slack-market/api/endpoints/slack_endpoint.rb

```diff
@@ -0,0 +1,42 @@
+require 'pp'
+module Api
+  module Endpoints
+    class SlackEndpoint < Grape::API
+      format :json
+
+      namespace :graph do
+        desc 'Select graph.'
+        post do
+          # slack interactive messages send a payload.
+          # the code below formats that object into a slack message
+          # response to the channel
+          payload = JSON.parse(params[:payload])
+          button_name = payload['actions'][0]['name']
+          button_value = payload['actions'][0]['value']
+          stock_symbol = button_value.scan(/^[^\-]*/)
+          channel = payload['channel']['id']
+          token = payload['token']
+          ts = payload['original_message']['ts']
+          chart = true
+          quotes = Market.quotes([stock_symbol])
+
+          slack_attachment = Market.to_slack_attachment(quotes[0], chart, button_name)
+          # verifying message token
+          if token == ENV['SLACK_VERIFICATION_TOKEN']
+            # formatted Slack message response
+            {
+              as_user: true,
+              channel: channel,
+              ts: ts,
+              token: token,
+              text: "Below is the #{button_name} chart",
+              attachments: [slack_attachment]
+            }
+          else
+            fail 'Message token is not actually coming from Slack.'
```

Comment:
This should be a 401 via `error! ...`, see https://github.com/ruby-grape/grape#raising-exceptions. With `fail` it's a 500, the server has failed, but in this case the problem is the client's data, so we want a 401.